src/mavlink/mavlinkmanager.cpp

```diff
@@ -0,0 +1,80 @@
+#include "mavlinkmanager.h"
+
+#include "logger.h"
+
+#include <mavlink_msg_attitude.h>
+
+#include <QDebug>
+#include <QQmlEngine>
+
+PING_LOGGING_CATEGORY(MAVLINKMANAGER, "ping.mavlinkmanager")
+
+MavlinkManager::MavlinkManager()
+{
+    QQmlEngine::setObjectOwnership(this, QQmlEngine::CppOwnership);
+
+    // This is our default udpin connection link
+    // TODO: Allow the user to configure this
+    connect({LinkType::Udp, {"192.168.2.2", "14660"}});
+
+    QObject::connect(&_heartbeatTimer, &QTimer::timeout, this, &MavlinkManager::sendHeartbeatMessage);
+    _heartbeatTimer.start(4000);
+}
+
+void MavlinkManager::connect(const LinkConfiguration& conf)
+{
+    if (_linkIn) {
+        auto link = _linkIn->self();
+        if (link->isOpen()) {
+            link->finishConnection();
+        }
+    }
+
+    qCDebug(MAVLINKMANAGER) << "Connecting to" << conf;
+    if (!conf.isValid()) {
+        qCWarning(MAVLINKMANAGER) << LinkConfiguration::errorToString(conf.error());
+        return;
+    }
+    if (_linkIn) {
+        _linkIn.clear();
+    }
+    _linkIn = QSharedPointer<Link>(new Link(conf));
+    _linkIn->self()->startConnection();
+
+    QObject::connect(_linkIn->self(), &AbstractLink::newData, this, &MavlinkManager::parseData);
+}
+
+void MavlinkManager::parseData(const QByteArray data)
+{
+    static mavlink_message_t message;
+    mavlink_status_t status;
+    for (const auto& byte : data) {
+        if (mavlink_parse_char(0, byte, &message, &status)) {
+            emit mavlinkMessage(message);
+        }
+    }
+}
+
+void MavlinkManager::sendHeartbeatMessage()
+{
+    static mavlink_message_t message;
+    static unsigned char bufffer[1024];
```

Comment:
I wonder if the `static thread mutex` that you discovered will be bad here. maybe those should be defined in an unamed namespace in the cpp context?

Reply:
That's a good idea, but I would prefer to initialize the message in compile time.

Reply:
Fixed.